kernel/typing.ml

```diff
@@ -399,3 +399,11 @@ let check_rule sg (rule:untyped_rule) : SS.t * typed_rule =
     pat = rule.pat;
     rhs = rule.rhs
   }
+
+let typed_rule_of_ri s ri =
+  let ur =
+    { name = ri.name
+    ; ctx  = infer_rule_context ri
+    ; pat  = pattern_of_rule_infos ri
+    ; rhs  = ri.rhs} in
+  snd (check_rule s ur)
```

Comment:
You should return also the substitution. The user will select if he needs it or not.